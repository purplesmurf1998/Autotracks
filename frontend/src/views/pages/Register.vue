<template>
  <div class="c-app d-flex align-items-center min-vh-100" :class="{ 'c-dark-theme': $store.state.darkMode }">
    <CContainer fluid>
      <CRow class="justify-content-center">
        <CCol md="6">
          <CButton color="secondary" class="mx-4 mb-2" @click="goBack">
            Back
          </CButton>
          <register-admin :adminUser.sync="adminUser"/>
          <CRow class="mt-2 d-flex justify-content-center">
            <CButton color="primary" @click="completeRegistration">
              Complete Registration
            </CButton>
          </CRow>
        </CCol>
      </CRow>
    </CContainer>
  </div>
</template>

<script>
import RegisterAdmin from "../../modules/RegisterAdmin.vue";

export default {
  name: 'Register',
  components: {
    'register-admin': RegisterAdmin,
  },
  data() {
    return {
      // register admin field values
      adminUser: {
        adminFirstName: '',
        adminLastName: '',
        adminEmail: '',
        adminPassword: '',
        adminConfirmPassword: ''
      },
    }
  },
  methods: {
    goBack () {
      this.$router.push('/pages/login');
    },
    completeRegistration () {
      this.$router.push('/dashboard');
    }
  }
}
</script>
