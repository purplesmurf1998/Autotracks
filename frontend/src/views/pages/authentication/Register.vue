<template>
  <div class="c-app d-flex align-items-center min-vh-100" :class="{ 'c-dark-theme': $store.state.darkMode }">
    <CContainer fluid>
      <CRow class="justify-content-center">
        <CCol md="6">
          <CButton color="secondary" class="mx-4 mb-2" @click="goBack">
            Back
          </CButton>
          <!-- Register admin component -->
          <CCard class="mx-4 mb-0">
            <CCardBody class="p-4">
              <CForm>
                <h1>Step 1: Register new account!</h1>
                <p class="text-muted">This account will be the admin account to the dealership.</p>
                <CRow>
                  <CCol>
                    <CInput
                      id="first-name-txt-field"
                      label="First Name"
                      placeholder="ex. John"
                      autocomplete="username"
                      v-model="adminUser.adminFirstName"
                    />
                  </CCol>
                  <CCol>
                    <CInput
                      id="last-name-txt-field"
                      label="Last Name"
                      placeholder="ex. Doe"
                      autocomplete="username"
                      v-model="adminUser.adminLastName"
                    />
                  </CCol>
                </CRow>
                <CInput
                  id="email-txt-field"
                  label="email"
                  placeholder="ex. john.doe@mail.com"
                  autocomplete="email"
                  v-model="adminUser.adminEmail"
                />
                <CInput
                  id="password-txt-field"
                  label="Password"
                  placeholder="password"
                  type="password"
                  autocomplete="new-password"
                  v-model="adminUser.adminPassword"
                />
                <CInput
                  id="confirm-password-txt-field"
                  label="Confirm Password"
                  placeholder="confirm password"
                  type="password"
                  autocomplete="new-password"
                  v-model="adminUser.adminConfirmPassword"
                />
                <CInput
                  label="Account Role"
                  value="Administration"
                  disabled
                />
              </CForm>
            </CCardBody>
          </CCard>
          <CAlert show color="danger" v-if="errorMessage" class="mt-2">Invalid Credentials</CAlert>
          <CRow class="mt-2 d-flex justify-content-center">
            <CButton color="primary" @click="completeRegistration">
              Complete Registration
            </CButton>
          </CRow>
        </CCol>
      </CRow>
    </CContainer>
  </div>
</template>

<script>
const axios = require('axios');

export default {
  name: 'Register',
  data() {
    return {
      // register admin field values
      adminUser: {
        adminFirstName: '',
        adminLastName: '',
        adminEmail: '',
        adminPassword: '',
        adminConfirmPassword: ''
      },
      errorMessage: null
    }
  },
  methods: {
    goBack () {
      this.$router.push('/pages/login');
    },
    async completeRegistration () {
      /**********************************************
       * used axios since from the documentation axios() and fetch() are the most
       * commonly used http request APIs for SPAs.
       *
       * I left the code used with got in case you wanted to fix it but I was getting
       * an undefined error and couldn't figure out how to fix it. I think a problem
       * with the import.
       *
       * TODO: add some validation check (password = confirmPassword) and visual feedback
       *       when a field is entered wrong or not entered at all.
      ***********************************************/

      // send request to register new user
      axios({
        method: 'POST',
        url: 'http://localhost:5000/api/v1/auth/register',
        data: {
          first_name: this.adminUser.adminFirstName,
          last_name: this.adminUser.adminLastName,
          email: this.adminUser.adminEmail,
          role: "Administration",
          permissions: [
            "Add Dealerships",
            "View Dealerships",
            "Edit Dealerships",
            "Delete Dealerships",
            "Add Staff Users",
            "View Staff Users",
            "Edit Staff Users",
            "Delete Staff Users",
            "Add Vehicles",
            "View Vehicles",
            "Edit Vehicle Properties",
            "Edit Vehicle Locations",
            "Sell Vehicles",
            "Delete Vehicles"
          ],
          password: this.adminUser.adminPassword
        }
      }).then(async (response) => {
        console.log(response);
        //double check valid response
        if (response.status == 200) {
          // log the user in
          const response = await this.$store.dispatch('login', {
            email: this.adminUser.adminEmail,
            password: this.adminUser.adminPassword
          });

          // if not successful, show the error message
          if (!response.success) {
            this.showErrorMessage(response.message);
          }
          // if successful, redirect the user to the dashboad
          else {
            this.$router.push("/dashboard");
          }
        }
      }).catch(err => {
        console.log(err);
        // send an invalid registration message
        this.showErrorMessage('Invalid information or duplicate account.');
        this.adminUser.adminPassword = '';
        this.adminUser.adminConfirmPassword = '';
      })
    },
    showErrorMessage(msg) {
      // show the error message for 5 seconds and then dissapear
      this.errorMessage = msg;
      setTimeout(() => {
        this.errorMessage = null;
      }, 5000)
    }
  }
}
</script>
