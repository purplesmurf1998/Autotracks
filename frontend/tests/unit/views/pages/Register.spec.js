import {createLocalVue, shallowMount} from '@vue/test-utils'
import CoreuiVue from '@coreui/vue-pro'
import Vuex from 'vuex'

import Register from './../../../../src/views/pages/authentication/Register.vue'

const localVue = createLocalVue()
localVue.use(CoreuiVue)
localVue.use(Vuex)

const store = new Vuex.Store({
  state: {
    sidebarShow: 'responsive',
    sidebarMinimize: false,
    asideShow: false,
    darkMode: false
  }
})

describe('Register.vue', () => {
  it('has a name', () => {
    expect(Register.name).toBe('Register')
  })
  it('is Vue instance', () => {
    const wrapper = shallowMount(Register, { store, localVue })
    expect(wrapper.vm).toBeTruthy()
  })
  it('is Register', () => {
    const wrapper = shallowMount(Register, { store, localVue })
    expect(wrapper.findComponent(Register)).toBeTruthy()
  })
  it('should render correct content', () => {
    const wrapper = shallowMount(Register, { store, localVue })
    expect(wrapper.find('h1').text()).toMatch('Step 1: Register new account!')
    expect(wrapper.find('p').text()).toMatch('This account will be the admin account to the dealership.')
    expect(wrapper.find('#email-txt-field').exists())
    expect(wrapper.find('#password-txt-field').exists())
    expect(wrapper.find('#email-txt-field').exists())
    expect(wrapper.find('#login-btn').exists())
    expect(wrapper.find('#forgot-pswrd-btn').exists())
  })
  test('renders correctly', () => {
    const wrapper = shallowMount(Register, { store, localVue })
    expect(wrapper.element).toMatchSnapshot()
  })
})
