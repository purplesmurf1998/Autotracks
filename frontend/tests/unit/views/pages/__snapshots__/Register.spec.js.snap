// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register.vue renders correctly 1`] = `
<div
  class="c-app d-flex align-items-center min-vh-100"
>
  <ccontainer-stub
    fluid="true"
    tag="div"
  >
    <crow-stub
      class="justify-content-center"
      gutters="true"
      tag="div"
    >
      <ccol-stub
        md="6"
        tag="div"
      >
        <cbutton-stub
          activeclass="router-link-active"
          class="mx-4 mb-2"
          color="secondary"
          event="click"
          exactactiveclass="router-link-exact-active"
          id="registration-back-btn"
          routertag="a"
          target="_self"
          type="button"
        >
          
          Back
        
        </cbutton-stub>
         
        <ccard-stub
          class="mx-4 mb-0"
        >
          <ccardbody-stub
            class="p-4"
          >
            <cform-stub>
              <h1>
                Step 1: Register new account!
              </h1>
               
              <p
                class="text-muted"
              >
                
                This account will be the admin account to the dealership.
              
              </p>
               
              <crow-stub
                gutters="true"
                tag="div"
              >
                <ccol-stub
                  tag="div"
                >
                  <cinput-stub
                    autocomplete="username"
                    id="first-name-txt-field"
                    label="First Name"
                    lazy="400"
                    placeholder="ex. John"
                    type="text"
                    value=""
                  />
                </ccol-stub>
                 
                <ccol-stub
                  tag="div"
                >
                  <cinput-stub
                    autocomplete="username"
                    id="last-name-txt-field"
                    label="Last Name"
                    lazy="400"
                    placeholder="ex. Doe"
                    type="text"
                    value=""
                  />
                </ccol-stub>
              </crow-stub>
               
              <cinput-stub
                autocomplete="email"
                id="email-txt-field"
                label="email"
                lazy="400"
                placeholder="ex. john.doe@mail.com"
                type="text"
                value=""
              />
               
              <cinput-stub
                autocomplete="new-password"
                id="password-txt-field"
                label="Password"
                lazy="400"
                placeholder="password"
                type="password"
                value=""
              />
               
              <cinput-stub
                autocomplete="new-password"
                id="confirm-password-txt-field"
                label="Confirm Password"
                lazy="400"
                placeholder="confirm password"
                type="password"
                value=""
              />
               
              <cinput-stub
                disabled=""
                label="Account Role"
                lazy="400"
                type="text"
                value="Administration"
              />
            </cform-stub>
          </ccardbody-stub>
        </ccard-stub>
         
        <!---->
         
        <crow-stub
          class="mt-2 d-flex justify-content-center"
          gutters="true"
          tag="div"
        >
          <cbutton-stub
            activeclass="router-link-active"
            color="primary"
            event="click"
            exactactiveclass="router-link-exact-active"
            id="complete-registration-btn"
            routertag="a"
            target="_self"
            type="button"
          >
            
            Complete Registration
          
          </cbutton-stub>
        </crow-stub>
      </ccol-stub>
    </crow-stub>
  </ccontainer-stub>
</div>
`;
